### PR TITLE
First steps of BG2 EE launchability

### PR DESCRIPTION
**Pull request by MarcelHB**
## Description
Tiny Christmas project accomplished:

gemrb_bg2ee

So what was my goal:
* Check what is needed to launch an EE game without crashes
* Play the intro videos
* Reach the very first menu
* Provide a base to actually  allow someone commencing with the rest (probably all the GUI stuff)

What did I do exactly so far:
* Installed BG2EE -- this is probably the most advanced and feature-enriched version of all the EEs, so I gave BG2 precedence over BG/IWD/PST (I have them in GOG, so that's not a problem anyway) for the first throw
* Added some configs for the path to movies and the multi-language deployment (defaults to `lang/en_US`)
* Copied some of the unhardcoded stuff from BG2 into BG2EE and added `bg2ee` as a game switch
* Checked that VLC is doing a proper playback of the WEBM videos
* Added support for PVRZ DXT1 & DXT5 as far as what I found necessary until now (#1099)
* Added v2 support to BAM & MOS to load stuff from PVRZ files
* Wrote a `bg2ee/Start.py` that plays intro videos, the main music, and loads the first screen (screenshot)

This is really designed as the bare minimum so that _we can_ by now. Actually managing all the other major or minor details in #164, and doing all the GUIScript work is something for the future.

Also doing this DXT1/5 work in software is both unaccelerated and little-to-not portable through SDL2 r/n, and probably hurts the memory a little more. This is also open for optimization, but maybe also requires some other rework.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] I used the same coding style as the surrounding code 
- [x] I have tested the proposed changes
- [x] I extended the documentation, if necessary
- [ ] The proposed change builds also on our build bots (check after submission)

### CODE REVIEW

**Comment by MarcelHB** [comment]
I will look at the GCC error later, mine is much newer.

**Comment by lynxlynxlynx** [comment]
Oh wow, very nice! Especially how you were able to make BAM fit into the existing AnimationMgr. I thought the lack of FLT data would require more hacks.

And a bit of misfortune — I also started working on PVR/BAM2/TIS2/MOS2 support before NY, but got distracted by some IRL emergencies. I've added the intended BAM to the demo (original cave entrance swirl), so this will be easy to test without EEs. TIS has a bit shorter page + different naming convention + -1 page means full black. MOS and BAM share the same format, but they're the only ones so duplication is fine.

For the actual decoding I was considering mostly two:
- ffmpegs looks very intertwined with the rest of the code, didn't bother
- libsquish has a nice API, available on brew, but also quite big in comparison, so didn't bother
- bcdec, which can nicely be header-only, with optimized memory print and algorithmic optimizations
- openS3TC, which claims to be the fastest, but bcdec didn't exist yet and both use only software optimizations

I've stripped bcdec of the formats we don't need, while openS3TC had them per-file. I can push all my unobviated work to a branch, if you think it'd be useful.

**Comment by MarcelHB** [comment]
I hope I did not steal too much time then. Then again, since I did the DXT-stuff on my own for the most of the time here: The decoding itself is not that of a big thing, it was more difficult to deal with the non-aligned access and keeping the copying reasonable.

If any of these libraries had found SIMD optimiziation for decoding I'd go for it, too, but the three libs that I quickly scanned did not appear to have that. So I looked at the NI source code and did that little puzzling for fun.

Since you mentioned the TIS format, do you already have completed hat? I can offer to cherry-pick and wire this and your other modifications into this branch.

**Comment by lynxlynxlynx** [comment]
No worries, I suck at gfx, so it was a learning challenge as well and that part can't be discounted.

TIS I haven't done yet, just documented the mentioned data structure differences, the main being:
```c++
struct PVRDataBlock {
       // pageRef:
       // - MOS2, BAM2: refers to MOSxxxx.PVRZ files, where xxxx is a zero-padded four-digits decimal number
       // - TIS2: AR2600N.TIS would refer to A2600Nxx.PVRZ
       //   - the first character of the TIS filename
       //   - the four digits of the area code, the optional 'N' from night tilesets
       //   - this page value as a zero-padded two digits number
```

I get a crash with the BAM2 I created via NI though:
```gdb
Thread 1 "gemrb" received signal SIGSEGV, Segmentation fault.
0x00007ffff74c879f in unlink_chunk (p=p@entry=0x5555572ee640, av=0x7ffff7640c80 <main_arena>) at ./malloc/malloc.c:1628
1628    ./malloc/malloc.c: No such file or directory.
(gdb) bt
#0  0x00007ffff74c879f in unlink_chunk (p=p@entry=0x5555572ee640, av=0x7ffff7640c80 <main_arena>)
    at ./malloc/malloc.c:1628
#1  0x00007ffff74cb6ab in _int_malloc (av=av@entry=0x7ffff7640c80 <main_arena>, bytes=bytes@entry=200)
    at ./malloc/malloc.c:4307
#2  0x00007ffff74cc1b9 in __GI___libc_malloc (bytes=200) at ./malloc/malloc.c:3329
#3  0x00007ffff771da0c in operator new(unsigned long) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff5b10455 in GemRB::SDLVideoDriver::CreateSprite (this=0x5555555b9190, rgn=..., 
    pixels=0x555557431720, fmt=...) at /home/kajak/games/gemrb/gemrb/plugins/SDLVideo/SDLVideo.cpp:248
#5  0x00007ffff4a3e10c in GemRB::PVRZImporter::getSprite2DDXT5 (this=0x5555572a1ed0, region=...)
    at /home/kajak/games/gemrb/gemrb/plugins/PVRZImporter/PVRZImporter.cpp:381
#6  0x00007ffff4a3c729 in GemRB::PVRZImporter::GetSprite2D (this=0x5555572a1ed0, region=...)
    at /home/kajak/games/gemrb/gemrb/plugins/PVRZImporter/PVRZImporter.cpp:137
#7  0x00007ffff4a7fad6 in GemRB::BAMImporter::Blit (this=0x555556ac59d0, frame=..., dataBlock=..., 
    frameData=0x5555573f1710 "") at /home/kajak/games/gemrb/gemrb/plugins/BAMImporter/BAMImporter.cpp:214
#8  0x00007ffff4a7f7dd in GemRB::BAMImporter::GetV2Frame (this=0x555556ac59d0, frame=...)
    at /home/kajak/games/gemrb/gemrb/plugins/BAMImporter/BAMImporter.cpp:195
#9  0x00007ffff4a802dc in GemRB::BAMImporter::GetAnimationFactory (this=0x555556ac59d0, resref=..., 
    allowCompression=true) at /home/kajak/games/gemrb/gemrb/plugins/BAMImporter/BAMImporter.cpp:285
#10 0x00007ffff7c71929 in GemRB::GameData::GetFactoryResource (this=0x5555555affb0, resName=..., type=1000, 
    silent=false) at /home/kajak/games/gemrb/gemrb/core/GameData.cpp:392
(gdb) f 5
#5  0x00007ffff4a3e10c in GemRB::PVRZImporter::getSprite2DDXT5 (this=0x5555572a1ed0, region=...)
    at /home/kajak/games/gemrb/gemrb/plugins/PVRZImporter/PVRZImporter.cpp:381
381             auto spr = core->GetVideoDriver()->CreateSprite(Region{0, 0, region.w, region.h}, uncompressedData, fmt);
(gdb) p region
$1 = (GemRB::Region &&) @0x7fffffffca00: {origin = {x = 261, y = 1}, size = {w = 256, h = 256}, 
  x = @0x7fffffffca00, y = @0x7fffffffca04, w = @0x7fffffffca08, h = @0x7fffffffca0c}
```
It's using several regions in the same texture, but that should already be handled by the source stuff. I can push the files to master if you'd like a test case.

**Comment by MarcelHB** [comment]
Please do this, I will try to replay this then.

File: gemrb/GUIScripts/bg2ee/Start.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+def OnLoad():
+	global skip_videos
+
```
Comment (lynxlynxlynx) [inline gemrb/GUIScripts/bg2ee/Start.py:20]: I know it's a copy, but the global isn't needed. We can just replace the one use with the direct call to GetVar.

File: gemrb/GemRB.cfg.sample.in [hunk 2]
```diff
@@ -29,6 +29,7 @@
 #                                                   #
 #  bg1       Baldur's Gate                          #
 #  bg2       Baldur's Gate 2 : SoA or ToB           #
+#  bg2ee     Baldur's Gate 2 : EE (experimental!)   #
```
Comment (lynxlynxlynx) [inline gemrb/GemRB.cfg.sample.in:32]: You didn't add any GameType detection files, but I think it's fine and should stay that way for a while to not give false hope.

File: gemrb/core/Interface.cpp [hunk 3]
```diff
@@ -880,6 +880,8 @@ int Interface::Init(const InterfaceConfig* cfg)
 	CONFIG_PATH("GamePortraitsPath", config.GamePortraitsPath, "portraits");
 	CONFIG_PATH("GameScriptsPath", config.GameScriptsPath, "scripts");
 	CONFIG_PATH("GameSoundsPath", config.GameSoundsPath, "sounds");
+	CONFIG_PATH("GameLanguagePath", config.GameLanguagePath, "lang/en_US");
+	CONFIG_PATH("GameMoviesPath", config.GameMoviesPath, "movies");
```
Comment (lynxlynxlynx) [inline gemrb/core/Interface.cpp:884]: TODO: document in config samples and man page.

File: gemrb/plugins/BAMImporter/BAMImporter.cpp [hunk 4]
```diff
@@ -151,22 +252,41 @@ std::vector<BAMImporter::index_t> BAMImporter::CacheFLT()
...
+		for (const auto& frameInfo : frames) {
+			bool RLECompressed = allowCompression && frameInfo.RLE;
+			animframes.push_back(GetFrameInternal(frameInfo, RLECompressed, data - DataStart));
+		}
+		free(data);
+
+		return new AnimationFactory(resref, std::move(animframes), cycles, std::move(FLT));
+	} else {
+		std::vector<index_t> FLT;
+		FLT.resize(frames.size());
```
Comment (lynxlynxlynx) [inline gemrb/plugins/BAMImporter/BAMImporter.cpp:275]: Same as `std::vector<index_t> FLT(frames.size());` and the commented code below should be removed.

File: gemrb/core/Interface.cpp (hunk 3 above)
Comment (MarcelHB) [reply gemrb/core/Interface.cpp:884]: Given that we want to do it like this for now? I was just not sure if another EE or distribution requires a more comprehensive approach or something.

File: gemrb/unhardcoded/bg2ee/gemrb.ini [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+Palette256 = MPAL256
+
+; Length of a single step
+StepTime = 566
+
+; Base speed of chapter text scrolling
+TextScreenSpeed = 175
+
+; Default window title if none was specified by the user via GameName
+WindowTitle = Baldur's Gate 2
```
Comment (lynxlynxlynx) [inline gemrb/unhardcoded/bg2ee/gemrb.ini:67]: EE

File: gemrb/plugins/PVRZImporter/PVRZImporter.cpp [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+bool PVRZImporter::Import(DataStream* str) {
+	ieDword signature;
+	bool decompressed = false;
+
+	while (true) {
+		str->ReadDword(signature);
+
+		if (signature == 0x50565203) {
+			str->SetBigEndianness(true);
+		} else if (signature != 0x03525650) {
```
Comment (lynxlynxlynx) [inline gemrb/plugins/PVRZImporter/PVRZImporter.cpp:39]: Curiously this stands for PVR\3, not PVR3.

File: gemrb/core/Interface.cpp (hunk 3 above)
Comment (lynxlynxlynx) [reply gemrb/core/Interface.cpp:884]: The EEs are now all on the same engine, except for PST:EE, so there's not as much diversity as you think. We don't need to support older EE versions, but in this case I don't think it even matters, the layout hasn't changed from the start AFAIK.

File: gemrb/GemRB.cfg.sample.in (hunk 2 above)
Comment (MarcelHB) [reply gemrb/GemRB.cfg.sample.in:32]: I'd like to skip that for now since I don't know exactly what's qualifying here.

Comment (lynxlynxlynx) [reply gemrb/GemRB.cfg.sample.in:32]: WeiDU has a list we can reuse later.

File: gemrb/plugins/BAMImporter/BAMImporter.cpp (hunk 4 above)
Comment (MarcelHB) [reply gemrb/plugins/BAMImporter/BAMImporter.cpp:275]: Force of habit, as without caution the ctor `std::vector` may interpret the size as an item to incorporate.

**Comment by MarcelHB** [comment]
The problem with the crash is not understanding the frame's origin coordinates correctly? Not sure r/n. I appears to have been `(0, 0)` for all what I've seen so far so that did not matter much, but for the demo we have one `(128, 128)` on a frame with dimension `(256, 256)` and a datablock of dimension `(256, 256)`.

So I'm currently trying to place that sprite at `(128 + 256, 128 + 256)` which is overflowing. What brings me to the question: How to understand this `origin`? Or is this coordinate correct and we have to calculate the maximum possible dimension first?

**Comment by bradallred** [comment]
I don't have anything useful to say, just wanted to chime in to say this is impressive. Good work!

**Comment by MarcelHB** [comment]
Thanks :)

So this `origin` is irrelevant for the target buffer location. Yet there is something later going on around map animations, will report back when done.

**Comment by MarcelHB** [comment]
Here we go:

Peek 2023-01-07 20-43

File: gemrb/GemRB.cfg.noinstall.sample [hunk 7]
```diff
@@ -267,6 +268,30 @@ PluginsPath=./gemrb/plugins/
 #  Input Parameters                                 #
 #####################################################
 
+#####################################################
```
Comment (lynxlynxlynx) [inline gemrb/GemRB.cfg.noinstall.sample:271]: Should all be a few lines higher, so you don't split the input section into two.

**Comment by lynxlynxlynx** [comment]
Looks great, but I see from its position we ignore the BAM frame center when drawing it. I guess something like using `frame.bounds` directly instead of a new `region` instead.

File: gemrb/GemRB.cfg.noinstall.sample (hunk 7 above)
Comment (MarcelHB) [reply gemrb/GemRB.cfg.noinstall.sample]: C&P is hard sometimes ...

**Comment by MarcelHB** [comment]
Ah, that's the missing piece ... I was wondering what the reason was to have x, y there.

**Comment by lynxlynxlynx** [comment]
Do movies look ok for you? I get 4 compressed copies at once at 1024x786:
<details>
image
</details>

**Comment by MarcelHB** [comment]
Odd, no. For me, they look good at any resolution. Can your VLC version play them correctly when played independently? Given that the player and library version match.

**Comment by lynxlynxlynx** [comment]
They can, yes. It's also true the versioning doesn't match. Don't worry about it.
I've pushed the two other decoder libs here: https://github.com/gemrb/gemrb/commit/33e381f69e08937d72efde61cf0c5e5ba3875270

**Comment by lynxlynxlynx** [comment]
I'll just merge this important work; we can do TIS2 and other improvements later. TIS won't matter until we try to load an area.

**Comment by MarcelHB** [comment]
Btw my DXT-implementation appears to have a little advantage that these two don't provide: slices. They both are going for the whole texture.

If one needs the complete thing, this is good. But looking at the assets, 1. some of them have maybe up to 30% empty space and 2. some appear to contain unrelated GUI stuff (like `mos0181`), so this could turn out to be better w. r. t. overhead.

**Comment by lynxlynxlynx** [comment]
I'm not sure that's true for bcdec, since you have to invoke it on each pixel block yourself.

**Comment by MarcelHB** [comment]
You are right. But that also then leaves you to handle all this non-aligned pixel access/partial blocks afterwards, so the gain is probably limited concerning the amount of code that is required.

**Comment by lynxlynxlynx** [comment]
I don't care enough to benchmark it. We can worry about speed when it becomes a problem.

**Comment by MarcelHB** [comment]
Ok, I was even more concerned that my whole code was looking like a mess from the outside. :wink:

**Comment by lynxlynxlynx** [comment]
Now tob saves load and I was able to do the talking, first fight and pocket plane cutscene. Clearly TIS support is missing, but now we have a simple test case. I was very surprised that so much of the GUI is intact — it's not clear yet if a separate game type really makes sense, since it looks like we'd be copying most of the code&data (probably should be something like iwd vs how, same code, different unhardcoded data). But that's for some later future, there's plenty of other less obvious things to do that should also help the originals.

**Comment by MarcelHB** [comment]
Concerning the TIS support, do you want me to do this next? If so, I can probably do something within the next two or so. Otherwise, feel free.

**Comment by lynxlynxlynx** [comment]
Ok, I'll give it a shot. Feel free to work on something else. :)

**Comment by lynxlynxlynx** [comment]
Looks like an impressionist painting, i.e. worse than the original, but it works. I knew the new format was worse in some circumstances, but this much is unexpected. And I doubt postprocessing could fix it.

**Comment by lynxlynxlynx** [comment]
The area background, eg.:

**Comment by MarcelHB** [comment]
I suspect it's related to TIS tiles being DXT1 in contrast to DXT5 of the GUI, I'll report back later.

**Pull request merged**